Step3 PR 요청 드립니다.

vvsungho opened the pull request:
안녕하세요 : )
Step3 로또게임 자동 기능 구현 PR요청 및 리뷰 부탁드립니다.

기능 구현 중 2가지 궁금증이 있어 같이 질문드립니다.

1. 일급컬렉션에서는 final static 변수를 선언해도 괜찮을까요?
  ```java
  public class LottoPurchase {
        private final static int LOTTO_PURCHASE_PRICE_VALUE = 1000;
        private final List<LottoNumber> lottoNumbers;
  }
  ```
해당 경우에서 클래스 내에서 LOTTO_PURCHASE_PRICE_VALUE 변수와 lottoNumbers 변수를 같이 선언해도 되는지 질문드립니다 !

2. 메서드 네이밍 고민
getter처럼 값을 가져오는 메서드는 명사구, 동사구 중 어떤 방법이 적합한지 궁금합니다.

이번 리뷰도 잘부탁드립니다 !!

vvsungho committed:
test: 로또게임 숫자 빈값, null값 확인 테스트케이스 작성

vvsungho committed:
feat: 숫자합 계산 Production Code 구현

vvsungho committed:
test: 입력값에 대한 테스트케이스 코드 작성

vvsungho committed:
fix: 테스트 케이스 fail 메서드 버그 수정

커스텀구분자 테스트 메서드 버그 수정

vvsungho committed:
refactor: 음수 입력값에 따른 예외클래스 수정

RunTimeException -> IllegalArgumentException

vvsungho committed:
rename: 메서드명 명사 -> 동사로 수정

vvsungho committed:
docs: 기능요구사항 작성

vvsungho committed:
refactor: 역할에 따른 클래스 분리

- 문자열을 분할
- 문자 숫자 타입 변경
- 숫자 합계

vvsungho committed:
rename: 테스트 메서드명 변경

vvsungho committed:
docs: 3단계 기능요구사항 작성

vvsungho committed:
test: 로또 번호 발급 테스트 케이스 작성

vvsungho committed:
docs: 기능요구사항 추가 작성

vvsungho committed:
test: 로또번호 오름차순 정렬 테스트 케이스 작성

vvsungho committed:
Merge branch 'vvsungho' of https://github.com/next-step/java-lotto-pro into step3

 Conflicts:
	docs/README.md

vvsungho committed:
rename: 클래스명, 메서드명 수정

클래스명 -> 명사구로 네이밍, 메서드명 -> 확실한 의미를 전달할 수 있도록 네이밍

vvsungho committed:
test: 로또번호, 당첨번호 일치갯수 확인 테스트 케이스 작성

vvsungho committed:
test: 구매한 로또의 갯수를 확인한다.

vvsungho committed:
fix: 테스트케이스 성공하도록 수정

vvsungho committed:
test: 로또수익률 확인 테스트 케이스 작성

vvsungho committed:
feat: 로또번호 자동발급 기능 구현

vvsungho committed:
rename: 패키지 이동

vvsungho committed:
refactor: 로또번호 자동발급 기능 리펙토링

vvsungho committed:
feat: 발급된 n개의 로또번호 오름차순 정렬 기능 적용

vvsungho committed:
feat: 로또번호 당첨 갯수 확인 기능 적용

vvsungho committed:
feat: 로또 수익률 계산 기능 적용

vvsungho committed:
rename: 메서드명 수정

vvsungho committed:
refactor: 미사용 메서드 삭제

vvsungho committed:
refactor: 로또번호 벨리데이션 메서드 적용

vvsungho committed:
feat: view 클래스 생성

seokhongkim commented on docs/README.md:
```diff
@@ -8,3 +8,11 @@
 ---
+
+## 로또 (자동)
+### 기능요구사항
```
기능 요구사항을 체크리스트로 잘 작성해주셨네요! 👍 

3단계 미션을 모두 구현하시지는 않은 것 같아요.
화면에 로또와 관련된 내용을 출력하고, 값을 입력 받아서 실제 실행 결과를 출력하는 것까지 추가로 구현 부탁드려요!

seokhongkim commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class LottoNumber {
```
`LottoNumber` 가 일급 콜렉션 역할과 숫자를 관리하는 역할을 다 하고 있네요.
1~45 숫자를 관리하는 `LottoNumber` 와 6개의 숫자만 포함될 수 있는 `LottoNumbers` 역할을 분리해보는 건 어떨까요?

seokhongkim commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,81 @@
+    private final List<Integer> numbers;
+
+    public LottoNumber() {
+        List<Integer> randomNumbers = new ArrayList<>();
```
아래까지 로직을 봤을 때 `randomNumbers` 는 shuffle 이 되기 전까지 `random` 하지 않은 것으로 보이네요.
`randomNumbers` 라는 이름에 대해 고민해보면 좋을 것 같아요!

seokhongkim commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,81 @@
+
+    public LottoNumber() {
+        List<Integer> randomNumbers = new ArrayList<>();
+        for (int i = LOTTO_NUMBER_MIN_VALUE; i <= LOTTO_NUMBER_MAX_VALUE; i++) {
```
`LottoNumber` 가 1~45 범위의 값만 가질 수 있다면, 매번 이렇게 리스트를 만드는 것보다 더 효율적으로 생성할 수 있는 방법을 고민해보면 좋겠네요!

seokhongkim commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,81 @@
+        this.numbers = issueLottoNumber(numbers);
+    }
+
+    private List<Integer> issueLottoNumber(List<Integer> numbers) {
```
임의의 숫자로 로또를 발행하는 것이 `LottoNumber` 의 역할이 맞을까요? 🤔

seokhongkim commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,81 @@
+                .collect(Collectors.toList());
+    }
+
+    public int winningCount(List<Integer> answerNumberList) {
```
로또 번호와 우승 번호의 자료 구조를 List 로 사용하신 이유가 있을까요?

seokhongkim commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,81 @@
+        return LOTTO_NUMBER_SIZE_VALUE;
+    }
+
+    public List<Integer> sortedLottoNumber(List<Integer> numbers) {
```
1. 메소드 내용만 봤을 때, 입력 받은 리스트를 정렬해주는 역할만 하는 것으로 보입니다. 이름을 고민해보면 좋을 것 같아요!
2. 숫자가 정렬되어야 한다는 것은 로또의 요구사항일까요? 화면에 보여질 때의 요구사항일까요? 🤔

seokhongkim commented on src/main/java/lotto/LottoPurchase.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private Integer calculateWinningLottoPrice(int winningCount) {
+        if (isEqualNumber(winningCount, LOTTO_ONE_RANK_COUNT_VALUE))
```
`LOTTO_ONE_RANK_COUNT_VALUE` 와 일치하면 `LOTTO_ONE_RANK_PROFIT_VALUE` 를 사용해야 한다는 것을 코드를 작성한 사람만 알 수 있을 것 같아요!
물론 문맥상 두 값이 매치된다고 생각하겠지만, 읽고 파악해야 하는 것보다 직접적인 연관 관계를 나타내줄 수 있는 방법은 없을까요?

seokhongkim commented on src/main/java/lotto/LottoPurchase.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private Integer calculateWinningLottoPrice(int winningCount) {
+        if (isEqualNumber(winningCount, LOTTO_ONE_RANK_COUNT_VALUE))
```
반복되는 if-return 구문을 개선할 수 있는 방법을 고민해보시면 좋을 것 같아요!
1~4등까지 루프를 돌면서 값이 매치하면 상금을 반환하는 형태가 될 것 같네요.

vvsungho replied:
코멘트 감사드립니다 !
로또 상금 관리 Enum을 만들어 중복코드를 개선하였습니다.

seokhongkim commented on src/main/java/lotto/LottoWinning.java:
```diff
@@ -0,0 +1,41 @@
+    private final static double ROUNDING_DIGIT_VALUE = 100d;
+
+    private final LottoPurchase lottoPurchase;
+    private final int purchasePrice;
```
각 필드는 무슨 목적으로 가지고 있는 데이터일까요?
중간 계산 과정에서만 사용되는 값이라면 지역 변수로 충분할 것 같아요!

vvsungho committed:
feat: Lotto 게임 실행 Application 생성

seokhongkim reviewed (changes requested):
성호님, 주말에 3단계 구현 열심히 해주셨네요! 👍 
코멘트 조금 남겼으니, 내용 확인 후 반영 부탁드립니다. 😀

문의주신 내용에 대해서는 아래와 같이 답변드립니다.
> 일급컬렉션에서는 final static 변수를 선언해도 괜찮을까요?

static final 의 목적이 상수를 관리하기 위한 목적이고, 일급 컬렉션에서도 유의미한 상수가 있다면 충분히 선언할 수 있다고 생각합니다.
static final 을 사용하시는 이유, 일급 컬렉션에서 필요하신 이유를 고민해보시면 도움이 될 것 같습니다.

> 메서드 네이밍 고민
getter처럼 값을 가져오는 메서드는 명사구, 동사구 중 어떤 방법이 적합한지 궁금합니다.
로또 번호가 담긴 ArrayList를 오름차순으로 정렬한 값을 리턴하는 메서드명을 sortedLottoNumber()으로 네이밍했는데
리뷰어님의 의견을 듣고싶습니다 !

성호님께서 사용하신 것처럼 `.sortedLottoNumbers()` 로 꺼내오는 것도 방법일 것 같아요.
`.asSortedList()` 정도로도 사용해볼 수 있지 않을까 싶습니다. 😀

vvsungho committed:
refactor: 도메인 역할 분리

vvsungho committed:
feat: 결과 view 클래스 구현

vvsungho commented:
코멘트 주신 내용을 바탕으로 각 도메인에 대한 역할을 다시 생각하고 설계를 다시하다보니 다시 리뷰요청드리기까지 시간이 오래 걸렸네요 🤔

너무 꼼꼼한 코멘트 다시 한 번 감사드립니다 !

seokhongkim commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,23 @@
+    private final List<LottoNumber> lottoNumbers;
+
+    public Lotto(List<Integer> lottoNumbers) {
+        if (lottoNumbers.size() != LOTTO_NUMBER_SIZE)
```
`lottoNumbers` 에 대해 `중복된 숫자가 들어있으면 안된다` 또는 `중복 제거 후 6개의 숫자여야 한다` 는 제약조건 검증이 필요할 것 같아요!

seokhongkim commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,34 @@
+    public LottoResult(Lottos lottos, Lotto answerLotto) {
+        lottoResultMap = new HashMap<>();
+
+        lottos.getLottos().forEach(lotto -> {
```
`lottos` 에게 당첨 번호를 주고 당첨 결과를 알려달라고 하면 어떨까요?
`LottoResult` 가 `Lottos`, `Lotto` 에 대해 가지고 있는 의존성을 제거할 수 있을 것 같네요!

seokhongkim commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -1,85 +1,32 @@
 
-    private final List<Integer> numbers;
+    private static final String ERROR_LOTTO_NUMBER_OUT_OF_RANGE_MESSAGE = "[ERROR] 로또 번호는 1 ~ 45 사이의 숫자로 구성되어야합니다.";
+    private final Integer number;
```
로또 번호는 1~45 사이의 값만 가질 수 있으니, 미리 객체를 만들어서 재사용하는 방법을 고민해보시면 좋을 것 같아요 😀

seokhongkim commented on src/main/java/lotto/service/LottoAutoIssuedServiceImpl.java:
```diff
@@ -0,0 +1,30 @@
+    public List<Integer> issueLottoNumber() {
+        List<Integer> lottoTargetNumbers = new ArrayList<>();
+        for (int i = LOTTO_NUMBER_MIN_VALUE; i <= LOTTO_NUMBER_MAX_VALUE; i++)
+            lottoTargetNumbers.add(i);
```
`LottoNumber` 에서 1~45 범위의 객체를 미리 가지고 있도록 수정한다면, 발급할 때 매번 컬렉션을 새로 만들 필요 없이 만들어져 있는 컬렉션을 복사해서 사용해볼 수 있겠네요!

seokhongkim commented on src/main/java/lotto/type/LottoRank.java:
```diff
@@ -0,0 +1,38 @@
+    SECOND(5, 1500000),
+    THIRD(4, 50000),
+    FOURTH(3, 5000),
+    FIFTH(2, 0),
```
`FIFTH`, `SIXTH` 와 `NONE` 의 차이는 무엇일까요?

seokhongkim commented on src/main/java/lotto/type/LottoRank.java:
```diff
@@ -0,0 +1,38 @@
+    SIXTH(1, 0),
+    NONE(0, 5000);
+
+    public static final String ERROR_INVALID_MATCHED_COUNT_MESSAGE = "[ERROR] 번호 일치 갯수가 유효하지 않습니다.";
```
사용하지 않는 코드인 것 같아요!

seokhongkim commented on src/main/java/lotto/config/AppConfig.java:
```diff
@@ -0,0 +1,20 @@
+import lotto.view.InputView;
+import lotto.view.ResultView;
+
+public class AppConfig {
```
스프링 빈 등록처럼 컨피그를 사용해보신 걸까요?
`AppConfig` 로 분리했을 때 얻을 수 있는 장점은 어떤 게 있을까요?
저는 현재 시점에 인터페이스, 설정 등으로 분리했을 때 얻을 수 있는 장점이 명확하지 않다면 오히려 코드가 복잡해질 수 있다고 생각합니다. 성호님은 어떻게 생각하시나요?

vvsungho replied:
이후 로또수동발급까지 고려하여 로또발급방식(자동, 수동)을 서비스로 관리하고자 하였는데.. 오히려 코드 복잡도가 높아진 것 같습니다.

vvsungho replied:
코드의 복잡성이 높아진 것 같아 AppConfig 클래스는 삭제하였습니다.
추후 로또 수동 기능까지 생각한다면 로또발급기능을 기존처럼 인터페이스로 구현할지, 혹은 인터페이스 말고 Lotto클래스에서 별도 메서드로 구현할지 고민이되네요.. ㅠㅠ

seokhongkim replied:
저는 어떤 방법이 좋을지 확신이 없을 때는, 우선 동작하도록 구현 후 리팩토링을 하며 구조를 정리하는 방법도 잘 맞았어요.
성호님은 고민되실 때 어떤 방법으로 접근하시는지 궁금하네요!

seokhongkim reviewed (changes requested):
성호님, 전에 드린 피드백을 잘 반영해주셨네요! 👍 

추가로 피드백을 조금 남겼습니다. 확인 부탁드립니다. 😀

vvsungho committed:
refactor: 로또번호 중복 추가 검증

vvsungho committed:
refactor: LottoResult 로또당첨 결과 도메인 의존성 제거

vvsungho committed:
refactor: 로또순위 관리 이넘, 미사용 코드 삭제

vvsungho committed:
refactor: Lotto 번호 객체 재사용

vvsungho committed:
refactor: AppConfig 설정 삭제

vvsungho committed:
feat: 로또번호 정답 갯수 확인 기능 구현

vvsungho committed:
test: 로또등수 확인 테스트코드 추가

vvsungho committed:
refactor: 로또결과 확인 담당 객체 lottos -> lotto로 변경

seokhongkim commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,28 @@
+    private final static int LOTTO_NUMBER_MIN_VALUE = 1;
+    private final static int LOTTO_NUMBER_MAX_VALUE = 45;
+
+    private final List<Integer> numbers;
```
`LottoNumber` 도메인의 역할이 `List<Integer> numbers` 를 가지고 있는 역할로 축소된 것 같아요.

tip: `LottoNumber` 를 미리 만들어서 재사용하는 방법은 아래 내용을 참고하셔서 구현해보시면 좋겠어요!
* `List<LottoNumber> lottoNumbers` 에 생성될 수 있는 `LottoNumber` 를 미리 만들어 놓음
  *  예시로 List 를 들었지만, 필요에 따라 적절한 컬렉션 사용
* 특정 번호로 `LottoNumber` 를 만들려고 할 때, 미리 생성되어 있는 객체를 찾아서 반환

vvsungho replied:
바쁘신 와중에 꼼꼼한 리뷰 감사드립니다 !

코멘트남겨주신 부분을 토대로 코드를 수정해보았습니다.

1. LottoNumber의 numbers는 항상 1~45의 값을 가지므로 재사용할 수 있도록 static변수로 수정하였습니다.
2. LottoNumber의 numbers는 중복된 값을 가지지 않으므로, 자료형을 List에서 Set으로 수정하였습니다.

해당 단계에서 너무 많은 리뷰요청을 드렸네요.
죄송스럽지만, 다시 한 번만 리뷰요청 드립니다.  🙏

seokhongkim reviewed (changes requested):
성호님, 대부분의 피드백을 잘 반영해주셨습니다. 👍
`LottoNumber` 를 재사용하기 위해 리팩토링 하신 부분에 대해 간단한 피드백이 하나 있어서, 이 부분만 추가로 확인 부탁드립니다. 🙏

vvsungho committed:
refactor: LottoNumber 객체 수정

- 로또 번호는 항상 1 ~ 45의 숫자를 가지므로 해당값을 미리 가지고 있도록 수정
- 중복이 없는 숫자이므로 자료형 List -> Set으로 변경

seokhongkim commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -1,22 +1,24 @@
     private final static int LOTTO_NUMBER_MAX_VALUE = 45;
 
-    private final List<Integer> numbers;
+    public final static Set<Integer> numbers = generateNumbers();
```
성호님 리뷰 여러번 요청하시는 건 괜찮습니다. 😀 
이 코드는 아래 가이드를 참고해서 조금 더 고민해보시면 좋을 것 같아요.

1번은 필수로 적용해주시면 좋을 것 같고, 2번은 참고하셔서 학습만 해보셔도 괜찮습니다!

1. `LottoNumber` 의 핵심 역할은 `int value` 값을 감싸서(필드로 가지고) 1-45 사이의 값만 사용할 수 있도록 제한하는 것이라고 생각합니다. 이를 통해 `LottoNumber` 를 사용하는 곳에서는 이 값이 항상 1-45 사이의 값이라고 생각할 수 있죠. 값의 범위가 변하더라도 `LottoNumber` 에서 알아서 관리해줄 것이라고 생각할 수 있습니다.
2. `LottoNumber` 에서 `List<LottoNumber> lottoNumbers` 를 미리 가지고 있는 건 일종의 캐시 개념입니다. 1~45 사이의 값만 생성될 수 있으므로, 미리 대상 객체를 생성해놓고 재사용 할 수 있는거죠. 이건 성능상의 이점을 얻기 위한 부분입니다. 꼭 구현하지 않으셔도 괜찮습니다. 😀 

제가 정확히 가이드를 드리지 않아서 혼선이 있었던 것 같아요.
2번 내용은 무시하시고, 1번만 먼저 진행하신 후, 여력이 되시면 2번은 참고만 부탁드릴게요!

vvsungho replied:
안녕하세요 ! 피드백 주신 내용 토대로 코드를 다시 리팩토링 해보았습니다.

1. LottoNumber는 오직 로또번호숫자 범위를 관리할 수 있도록 수정하였습니다.
2. 로또를 발급하는 서비스에서 로또번호를 재사용할 수 있도록 처리하였습니다.

각 역할에 맞도록 도메인구현하는 것이 쉽지 않네요. 😂
리뷰 부탁드립니다 !

seokhongkim reviewed (changes requested):
성호님, `LottoNumber` 에 대해 코멘트 남겨주신 부분에 대해 답변과 추가 확인 요청드려요!
내용 보신 후 조금 더 자세한 가이드가 필요하시면 말씀 부탁드립니다. 🙏

vvsungho committed:
refactor: 도메인 리팩토링. (LottoNumber객체 수정)

LottoNumber가 로또번호 범위 (1~45)를 보장할 수 있도록 수정

seokhongkim commented on src/main/java/lotto/service/LottoAutoIssuedServiceImpl.java:
```diff
@@ -1,25 +1,29 @@
 
-    private static List<Integer> divideNumberList(List<Integer> lottoNumbers) {
-        return lottoNumbers.subList(0, LOTTO_NUMBER_SIZE_VALUE);
+    private static List<Integer> divideNumberList() {
```
다음 단계에서 메소드가 하는 역할을 조금 더 잘 서술할 수 있는 이름을 고민해보셔도 좋을 것 같네요! 😄

seokhongkim reviewed (approved):
성호님, 피드백 드린 내용 잘 반영해주셨어요! 👍 

3단계 PR은 여기서 머지하도록 하겠습니다.
3단계 너무 고생 많으셨고, 고민 많이 하신만큼 4단계에서는 더 잘 하실 수 있을거라고 기대하게 되네요.

4단계 미션도 잘 부탁드립니다. 🙏
